src/Authentication/Authentication/Common/GraphSession.cs

```diff
@@ -0,0 +1,128 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph.PowerShell.Authentication
+{
+    using System;
+    using System.Threading;
+    /// <summary>
+    /// The current <see cref="GraphSession"/>.
+    /// </summary>
+    public class GraphSession: IGraphSession
+    {
+        static GraphSession _instance;
+        static bool _initialized = false;
+        static ReaderWriterLockSlim sessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+        internal Guid _graphSessionId;
+
+        /// <summary>
+        /// Gets or Sets <see cref="IAuthContext"/>.
+        /// </summary>
+        public IAuthContext AuthContext { get; set; }
+        public static GraphSession Instance
+        {
+            get
+            {
+                try
+                {
+                    sessionLock.EnterReadLock();
+                    try
+                    {
+                        if (null == _instance)
+                        {
+                            throw new InvalidOperationException(ErrorConstants.Codes.SessionNotInitialized);
+                        }
+                        return _instance;
+                    }
+                    finally
+                    {
+                        sessionLock.ExitReadLock();
+                    }
+                }
+                catch (LockRecursionException lockException)
+                {
+                    throw new InvalidOperationException(ErrorConstants.Codes.SessionLockReadRecursion, lockException);
+                }
+                catch (ObjectDisposedException disposedException)
+                {
+                    throw new InvalidOperationException(ErrorConstants.Codes.SessionLockReadDisposed, disposedException);
+                }
+            }
+        }
+        public GraphSession()
+        {
+            _graphSessionId = Guid.NewGuid();
```

Comment:
> _graphSessionId = Guid.NewGuid(); [](start = 11, length = 34)

why not do this when you initialize and take away the ability to directly construct a graphsession all together.

Reply:
ok it's because of how you want to possibly initialize from a file in the future.

---
In reply to: [405665227](https://github.com/microsoftgraph/msgraph-sdk-powershell/pull/164#discussion_r405665227) [](ancestors = 405665227)

Reply:
That's right. We will use the initializer to read the auth context from a file and even support multiple login accounts.